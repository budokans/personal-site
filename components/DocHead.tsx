import Head from "next/head";
import { ReactElement } from "react";
import { SiteMetadata } from "types";

interface DocHeadProps {
  readonly metadata: SiteMetadata;
}

export const DocHead = ({ metadata }: DocHeadProps): ReactElement => (
  <Head>
    <title>{metadata.title}</title>
    <meta name="description" content={metadata.description} />

    <meta property="og:title" content={metadata.title} />
    <meta property="og:description" content={metadata.description} />
    <meta property="og:image" content={`${metadata.canonical}/og.png`} />
    <meta property="og:locale" content="en_US" />
    <meta property="og:type" content="website" />
    <meta property="og:site_name" content="Steven Webster" />

    <meta name="twitter:card" content="summary_large_image" />
    <meta name="twitter:title" content={metadata.title} />
    <meta name="twitter:image:src" content={`${metadata.canonical}/og.png`} />
    <meta name="twitter:description" content={metadata.description} />
    <meta name="twitter:url" content={metadata.canonical} />

    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
    <link rel="manifest" href="/site.webmanifest" />
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
  </Head>
);
