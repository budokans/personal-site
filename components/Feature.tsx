import { Box } from "@chakra-ui/layout";
import { MotionBox } from "./MotionBox";
import SimpleBar from 'simplebar-react';
import { CloseButton } from "@chakra-ui/close-button";
import { useFeatureContext } from "../lib/featureContext";
import { AnimatePresence } from "framer-motion";

type FeatureProps = {
  project: number,
}

const containerVariants = {
  hidden: {
    y: "100vh"
  },
  visible: {
    y: "5vh",
    transition: {
      type: "spring",
      mass: 0.3,
      damping: 8,
      when: "beforeChildren",
      staggerChildren: 0.4,
    },
  },
  exit: {
    y: "100vh",
    transition: {
      type: "spring",
      mass: 0.3,
      damping: 8,
      duration: 5
    }
  }
}

const Feature = () => {
  const {projectToFeature, showFeature, closeFeature} = useFeatureContext();

  return (
    <AnimatePresence>
      { showFeature && (
        <>
          <Box position="absolute" top={0} right={0} bottom={0} left={0} bg="gray.100" />
      
          <MotionBox position="fixed" left={0} right={0} bottom={0} maxWidth="930px" marginY="0" marginX="auto" height="100vh"  variants={containerVariants} initial="hidden" animate="visible" exit="exit">
            
            <Box height="full" bg="white" borderWidth="1px" borderColor="gray.300" borderStyle="solid" borderRadius="xl">
              
              <SimpleBar style={{ maxHeight: '100%' }}>
                <CloseButton size="md" fontSize="16px" borderRadius="full" color="gray.900" bg="gray.100" p={0} top="10px" right="10px" onClick={closeFeature}  />
              </SimpleBar>
            </Box>
          </MotionBox> 
        </>
      )}
    </AnimatePresence>
  )
}

export default Feature;
