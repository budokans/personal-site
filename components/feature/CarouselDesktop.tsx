import { Box } from "@chakra-ui/react";
import Image from "next/image";
import { hideScrollbar } from "../../styles/utilStyles";
import { MotionStack } from "../Motion";
import { FeatureMedia } from "../../interfaces";
import { ReactElement } from "react";

interface CarouselProps {
  readonly media: readonly FeatureMedia[];
}

export const CarouselDesktop = ({ media }: CarouselProps): ReactElement => {
  return (
    <MotionStack
      drag="x"
      dragConstraints={{ left: media.length === 5 ? -3020 : 0, right: 0 }}
      dragElastic={0.5}
      direction="row"
      alignItems="center"
      spacing={4}
      height="490px"
      whiteSpace="nowrap"
      sx={{ ...hideScrollbar, scrollSnapType: "x mandatory" }}
      cursor="grab"
    >
      {media.map((mediaItem, idx) => {
        return (
          <Box
            key={idx}
            border="1px solid black"
            bg={mediaItem.bg}
            borderRadius="2xl"
            height="100%"
            minWidth="763px"
            sx={{
              scrollSnapAlign: "start",
              scrollPadding: "1.75rem",
            }}
            p="55px"
          >
            <Box sx={{ pointerEvents: "none" }}>
              <Image
                src={mediaItem.path}
                alt={mediaItem.alt}
                height={378}
                width={651}
                priority={idx < 2}
                objectFit="cover"
              />
            </Box>
          </Box>
        );
      })}
    </MotionStack>
  );
};
