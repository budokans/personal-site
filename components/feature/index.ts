export * as Feature from "./Feature";
export * as Header from "./Header";
export * as Links from "./Links";
export * as Tech from "./Tech";
export * as Description from "./Description";
export { CarouselMobile } from "./CarouselMobile";
export { CarouselDesktop } from "./CarouselDesktop";
