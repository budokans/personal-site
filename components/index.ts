export { DocHead } from "./DocHead";
export * as Feature from "./feature";
export * as Home from "./home";
export * as Link from "./Link";
export * as Motion from "./Motion";
