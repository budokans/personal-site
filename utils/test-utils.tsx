import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
import { ChildrenProps } from "types";
import { ReactElement } from "react";

const withProviders = ({ children }: ChildrenProps): ReactElement => {
  return <ChakraProvider>{children}</ChakraProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: withProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
