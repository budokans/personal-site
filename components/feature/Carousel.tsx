import { ReactElement, useState, useCallback } from "react";
import { Box, Stack } from "@chakra-ui/react";
import Image from "next/image";
import { Motion } from "components";
import { FeatureMedia } from "types";
import { hideScrollbar } from "styles/utilStyles";

export interface CarouselProps {
  readonly media: readonly FeatureMedia[];
}

export const Desktop = ({ media }: CarouselProps): ReactElement => {
  const [offset, setOffset] = useState(0);

  const contentRef = useCallback((node: HTMLElement | null) => {
    node && setOffset(node.scrollWidth - node.clientWidth);
  }, []);

  return (
    <Motion.MotionStack
      drag="x"
      dragConstraints={{
        left: -offset,
        right: 0,
      }}
      direction="row"
      spacing={4}
      whiteSpace="nowrap"
      cursor="grab"
      whileTap={{ cursor: "grabbing" }}
      ref={contentRef}
    >
      {media.map((mediaItem, idx) => (
        <Box
          key={idx}
          border="1px solid"
          borderColor="gray.200"
          bg={mediaItem.bg}
          borderRadius="2xl"
          height="100%"
          width="85%"
          flexShrink={0}
          p="55px"
        >
          <CarouselImage mediaItem={mediaItem} idx={idx} />
        </Box>
      ))}
    </Motion.MotionStack>
  );
};

export const Mobile = ({ media }: CarouselProps): ReactElement => (
  <Stack
    direction="row"
    spacing={4}
    overflowX="scroll"
    sx={hideScrollbar}
    scrollSnapType="x mandatory"
    scrollPadding={4}
    pl={4}
    mx={-4}
  >
    {media.map((mediaItem, idx) => (
      <Box
        key={idx}
        bg={mediaItem.bg}
        borderRadius="xl"
        height="100%"
        width="85%"
        flexShrink={0}
        scrollSnapAlign="start"
        p="6vw"
        marginRight={idx === media.length - 1 ? "1rem !important" : "0"}
      >
        <CarouselImage mediaItem={mediaItem} />
      </Box>
    ))}
  </Stack>
);

interface CarouselImageProps {
  readonly mediaItem: FeatureMedia;
  readonly idx?: number;
}

const CarouselImage = ({ mediaItem }: CarouselImageProps): ReactElement => (
  <Image
    src={mediaItem.path}
    alt={mediaItem.alt}
    height={378}
    width={651}
    priority
    style={{
      objectFit: "cover",
      pointerEvents: "none",
    }}
  />
);
