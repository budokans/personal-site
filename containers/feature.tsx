import { useEffect } from "react";
import { Box, Divider } from "@chakra-ui/layout";
import parse from "html-react-parser";
import { ProjectInterface } from "../interfaces";
import { Feature } from "../components/feature/Feature";
import { Header } from "../components/feature/Header";
import { Tech } from "../components/feature/Tech";
import { Description } from "../components/feature/Description";
import { Links } from "../components/feature/Link";
import { CarouselContainer } from "../containers/carousel";

interface FeatureProps {
  project: ProjectInterface;
  onCloseClick: () => void;
}

export const FeatureContainer: React.FC<FeatureProps> = ({
  project,
  onCloseClick,
}) => {
  useEffect(() => {
    document.getElementsByTagName("body")[0].style.overflow = "hidden";
    return () => {
      document.getElementsByTagName("body")[0].style.overflow = "visible";
    };
  }, []);

  const containerVariants = {
    hidden: {
      height: "0%",
    },
    visible: {
      height: "95%",
      transition: {
        type: "spring",
        mass: 0.3,
        damping: 10,
      },
    },
    exit: {
      height: "0%",
      transition: {
        type: "spring",
        mass: 0.3,
        damping: 8,
      },
    },
  };

  return (
    <Feature>
      <Feature.Container
        variants={containerVariants}
        onCloseClick={onCloseClick}
      >
        <Feature.CloseButton onCloseClick={onCloseClick} />

        <Header>
          <Header.Image src={project.icon} alt={project.title} />
          <Header.Inner>
            <Header.Text>{project.title}</Header.Text>
            <Header.Subtext>{project.shortBlurb}</Header.Subtext>
          </Header.Inner>
        </Header>

        <Box py={4} px={[4, 9]}>
          <Divider orientation="horizontal" />
        </Box>

        <Feature.Inner>
          <Tech>
            <Tech.Header>Tech</Tech.Header>
            <Tech.Inner>
              {project.tech.map((tech, idx) => {
                return <Tech.Badge key={idx}>{tech}</Tech.Badge>;
              })}
            </Tech.Inner>
          </Tech>

          <CarouselContainer media={project.featureMedia} />

          <Description>
            {project.description.map((paragraph, idx) => {
              return (
                <Description.Paragraph key={idx}>
                  {parse(paragraph)}
                </Description.Paragraph>
              );
            })}
          </Description>

          <Links>
            {project.links.map((link) => {
              return (
                <Links.Link url={link.url} key={link.type}>
                  {link.type}
                </Links.Link>
              );
            })}
          </Links>
        </Feature.Inner>
      </Feature.Container>
    </Feature>
  );
};
