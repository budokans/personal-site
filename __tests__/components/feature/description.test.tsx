import { render, screen } from "test-utils";
import { Description } from "components/feature";

describe("<Description />", () => {
  test("renders a <Description.Paragraph> with populated data", () => {
    const { container } = render(
      <Description.Description>
        <Description.Paragraph>
          The foundations of the project were laid while completing Wes
          Bos&apos;s huge Advanced React with GraphQL course, but the final
          result was buggy and impossible to deploy in its finished state.
        </Description.Paragraph>
      </Description.Description>
    );
    expect(screen.getByText(/The foundations/)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
