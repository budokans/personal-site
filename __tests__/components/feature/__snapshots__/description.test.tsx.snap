// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Description /> renders a <Description.Paragraph> with populated data 1`] = `
<div
  class="css-1xnkfpe"
>
  <p
    class="chakra-text css-2256fn"
  >
    The foundations of the project were laid while completing Wes Bos's huge Advanced React with GraphQL course, but the final result was buggy and impossible to deploy in its finished state.
  </p>
</div>
`;
